### PR TITLE
Allow a minimum chunk size rather than enforcing a big one

### PR DESCRIPTION
**Pull request by jviotti**
Consider an image which size is 2621440 bytes, which equals to 2.5 MB.
This module enforces a block size of 1 MB, meaning that when processing
the last chunk (512 KB), it will get padded by another 512 KB.

This introduces the chance that the drive might not be able to hold the
padded image, but has enough space to hold to unpadded image, causing a
lot of confusion to the user.

In order to mitigate this problem, but still preserve the default big
block size (for speed reasons), we introduce the concept of a "minimum
block size". What this means is that the input stream is automatically
chunked to the default block size (e.g: 1 MB), but if a chunk is smaller
than the block size, its left alone _unless_ its smaller than the
minimum block size, in which case its padded to match the minimum block
size instead.

Basically, this means that 99% of the blocks will be padded to 1 MB, but
some of them, in rare ocasions where the image is not 1MB-aligned and
the drive is almost the same size as the image, they will be padded to
512 KB.

If a chunk size is bigger than the minimum block size, there is still a
chance that its very malformed, and could cause issues when being copied
to the actual drive. As workaround, blocks that are bigger that our
minimum block size get padded to match the initial block size.

See: https://github.com/resin-io/etcher/issues/685
Signed-off-by: Juan Cruz Viotti jviotti@openmailbox.org

### CODE REVIEW

**Comment by lurch** [comment]
@jviotti Over the years I was working on (and testing) NOOBS, I've amassed a large collection of (almost 100) different SD cards, the details of which I keep in a spreadsheet. The only SD cards which _don't_ have a size evenly divisible by 512KB are a "64 MB" card with 62128128 actual bytes (59.25 actual MB) and a "1 GB" card with 1022623744 actual bytes (975.25 actual MB).
Of the rest, 9 have a size evenly divisible by 512KB but not 1MB (2 of which are "32 GB" cards with capacities of 31914983424 bytes and 32009355264 bytes).
:nerd_face: 

So I agree that all modern cards will probably have a size evenly divisible by 512KB. The size of SD cards is relevant to Etcher because people often create "disk images" by using e.g. `dd` to back up the contents of an existing SD card, which they might later write (with Etcher) to a larger SD card.

**Comment by lurch** [comment]
> As far as I could see, Windows tends to throw EINVAL for
anything below 512kb.

Huh, so how does / would Win32DiskImager work in these cases? Are you sure the limit is 512kb and not 512b (which is the typical block size used by most disks and filesystems) ?

**Comment by jviotti** [comment]
You're right, I got confused, its 512b. I guess we could fallback to even smaller block sizes as we reach to the end of the image to cover all cases. What do you think?

**Comment by lurch** [comment]
I think it probably makes sense to leave things as-is until we get reports of people trying to use Etcher to flash disk images which have a size not evenly divisible by 512KB.
